compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/FieldVisibilityLowering.kt

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.jvm.lower
+
+import org.jetbrains.kotlin.backend.common.ClassLoweringPass
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.backend.jvm.JvmBackendContext
+import org.jetbrains.kotlin.descriptors.Visibilities
+import org.jetbrains.kotlin.ir.declarations.IrClass
+import org.jetbrains.kotlin.ir.declarations.IrDeclarationOrigin
+import org.jetbrains.kotlin.ir.declarations.IrProperty
+import org.jetbrains.kotlin.ir.util.hasAnnotation
+import org.jetbrains.kotlin.load.java.JvmAbi.JVM_FIELD_ANNOTATION_FQ_NAME
+
+internal val fieldVisibilityPhase = makeIrFilePhase(
+    ::FieldVisibilityLowering,
+    name = "FieldVisibilityLowering",
+    description = "Set correct visibility flags for fields."
+)
+
+private class FieldVisibilityLowering(val context: JvmBackendContext) : ClassLoweringPass {
+    override fun lower(irClass: IrClass) {
+        for (irProperty in irClass.declarations.filterIsInstance<IrProperty>()) {
+            if (irProperty.origin == IrDeclarationOrigin.FAKE_OVERRIDE) continue
+            val field = irProperty.backingField ?: continue
+            if (irProperty.isConst || field.hasAnnotation(JVM_FIELD_ANNOTATION_FQ_NAME)) continue
```

Comment:
I wonder if it makes more sense to implement most of the field visibility rules directly in psi2ir when creating the fields, so that all backends would observe the same IR structure and we'd have less trouble adapting common lowerings to be run on all backends. Namely, everything except the `@JvmField` logic here looks as if it should be performed by psi2ir. This lowering would still be needed, but only to change visibility of `@JvmField` fields from private to public. What do you think?

Reply:
I've updated the PR to make this change in psi2ir. Instead of adding a separate lowering for backend specific visibility exceptions, I've added a new function to `GeneratorExtensions` to compute field visibilities. This is then overridden in the jvm backend to handle `@JvmFields` and fields overridden from Java and in the js backend to handle `@JsExport`.

The reason is that we don't generate fake overrides for private fields (this would break during the IR verification phase in the JS IR backend), so a lowering to modify field visibility after the fact would have to add fake overrides as well. It's simpler to start out with the correct visibilities coming from psi2ir.

Unfortunately this affects a lot of IrTextTests. I've updated the tests in a separate commit.